### PR TITLE
Port remote configuration option to nginx.

### PR DESCRIPTION
**Pull request by crowell**
Ports config option and tests to nginx.

depends on the addition to psol first.

still tracking down crash when the remote server is on the same nginx instance as the server fetching the config. (this issue doesn't show up on apache)

### CODE REVIEW

**Comment by crowell** [comment]
the crash mentioned in the first comment was due to `worker_processes` being set to 1.

I talked to @jeffkaufman offline and am bumping the value to 2 for this test so that the test can be self-contained and not rely on a secondary server.

File: src/ngx_pagespeed.cc [hunk 1]
```diff
@@ -1265,6 +1265,22 @@ ps_loc_conf_t* ps_get_loc_config(ngx_http_request_t* r) {
...
+  if (!cfg_s || !cfg_s->server_context ||
+      !cfg_s->server_context->global_options()) {
+    return NULL;
+  }
+  if (!cfg_s->server_context->global_options()
+          ->remote_configuration_url()
+          .empty()) {
+    RewriteOptions* remote_options =
+        cfg_s->server_context->global_options()->Clone();
+    cfg_s->server_context->GetRemoteOptions(remote_options);
```
Comment (jeffkaufman) [inline src/ngx_pagespeed.cc]: Can you comment that this is a blocking fetch?  Things like that are very unusual here.

File: test/nginx_system_test.sh [hunk 2]
```diff
@@ -252,7 +252,7 @@ fi
 
 # check_stat in system_test_helpers.sh needs to know whether statstistics are
 # enabled, which is always the case for ngx_pagespeed.
-statistics_enabled=1
+statistics_enabled=0
```
Comment (jeffkaufman) [inline test/nginx_system_test.sh]: This isn't a good fix for the statistics issue.  As far as I know only the cache flushing test has a problem, but here we disable all stats testing in ngx_pagespeed.

File: test/pagespeed_test.conf.template [hunk 3]
```diff
@@ -1058,6 +1059,59 @@ http {
...
   }
 
+  # Start remote config servers.
+  server {
+    listen @@SECONDARY_PORT@@;
+    listen [::]:@@SECONDARY_PORT@@;
+    server_name remote-config.example.com;
+    pagespeed FileCachePath "@@SECONDARY_CACHE@@";
+    pagespeed RemoteConfigurationUrl "http://localhost:@@PRIMARY_PORT@@/mod_pagespeed_test/remote_config/remote.cfg?PageSpeed=off";
+    #pagespeed RemoteConfigurationUrl "http://raxcity.com/remote.cfg";
```
Comment (jeffkaufman) [inline test/pagespeed_test.conf.template]: remove before submitting

File: test/pagespeed_test.conf.template [hunk 4]
```diff
@@ -1058,6 +1059,59 @@ http {
...
+  }
+  server {
+    listen @@SECONDARY_PORT@@;
+    listen [::]:@@SECONDARY_PORT@@;
+    server_name remote-config-partially-invalid.example.com;
+    pagespeed FileCachePath "@@SECONDARY_CACHE@@";
+    pagespeed RemoteConfigurationUrl "http://localhost:@@PRIMARY_PORT@@/mod_pagespeed_test/remote_config/remote.cfg.partially.invalid";
```
Comment (jeffkaufman) [inline test/pagespeed_test.conf.template]: shouldn't this one and the ones below all have `?PageSpeed=off` ?

Comment (jeffkaufman) [inline test/pagespeed_test.conf.template]: Or is the exception in `location /mod_pagespeed_test/remote_config/` enough, in which case we can remove it from L1068 above?

Comment (crowell) [inline test/pagespeed_test.conf.template]: moved tests to use nc.

File: test/pagespeed_test.conf.template [hunk 5]
```diff
@@ -1078,6 +1078,57 @@ http {
     pagespeed DisableFilters remove_comments,add_instrumentation;
   }
 
+  # Start remote config servers.
+  server {
+    listen @@SECONDARY_PORT@@;
+    listen [::]:@@SECONDARY_PORT@@;
+    server_name remote-config.example.com;
+    pagespeed FileCachePath "@@SECONDARY_CACHE@@";
+    pagespeed RemoteConfigurationUrl "http://127.0.0.1:9991/remote.cfg";
```
Comment (jeffkaufman) [inline test/pagespeed_test.conf.template]: In the apache test you made these ports be substituted variables; want to do that here too?

**Comment by crowell** [comment]
$RCPORTN added.

**Comment by jeffkaufman** [comment]
LGTM.

Could you squash the commit history and force push?  There are a lot of commits in there that don't need to end up in trunk.

(Don't merge until the PSOL side of the change is in, though.)

**Comment by crowell** [comment]
ugh, broke it squashing, redoing in a minute

**Comment by crowell** [comment]
deleting this one, see #925 (same code pre breaking squash)

**Pull request closed without merging**